### ISSUE TITLE
Create relationship statement using elementId with v5 db is order of magnitude slower compared to using id in v4

### ISSUE THREAD

**Issue by bserdar** [neo4j/neo4j#13036]
- Neo4j Version: neo4j-community-5.4.0
 - Operating system: Fedora 37
 - Driver: Go 

I am incrementally inserting a graph into the Neo4j database. For this purpose, I load the existing graph, compute the delta, and then insert/update nodes/relationships. This is done in several stages within a transaction: first I insert the nodes, then using the node ids/element ids inserted in the first stage, I insert the relationships. For Neo4j version 4.4, the statement I am running is:

```
unwind $edges as edge 
match(a),(b) where id(a)=edge.from and id(b)=edge.to 
create (a)-[e:somelabel]->(b) set e=edge.props
```

Here, edge.from and edge.to are node ids inserted in the previous step.

This inserts 3000+ relationships in 90msecs.

When running against the 5.4.0 database, I use:

```
unwind $edges as edge 
match(a),(b) where elementId(a)=edge.from and elementId(b)=edge.to 
create (a)-[e:somelabel]->(b) set e=edge.props
```

The same 3000+ relationships insert in 15 seconds.

**Comment by HannesSandberg** [comment]
Thank you for your report, we will investigate this and come back to you.

**Comment by psytron** [comment]
```
WHERE split(elementId(n), ":")[2] = "0"
```
To get node by elementID we discard everything except the internal ID?  Whats the benefit of elementID?

**Comment by inmost-light** [comment]
@bserdar Thank you for reporting.

Could you try using `elementId(a) IN [edge.from] and elementId(b) IN [edge.to]` predicates and see if that makes a difference?

Also it would be useful if you could provide output from `EXPLAIN`.

**Comment by bserdar** [comment]
I finally had a chance to run some more diagnostics. The problem seems to be around returning the results.

This code is part of an open-source project we've been working on:

https://github.com/cloudprivacylabs/lsa-neo4j/blob/main/actions.go#L234

The (instrumented) code in question is something like this:
```
			start := time.Now()
			result, err := tx.Run(ctx, query, map[string]interface{}{"edges": unwind.u})
			if err != nil {
				return err
			}
			fmt.Println("Insert edges", len(unwind.u), time.Since(start), query)
			start = time.Now()
			records, err := result.Collect(ctx)
			if err != nil {
				return err
			}
			for i := range records {
				edge := unwind.insEdges[i]
				id := records[i].Values[0].(neo4j.Relationship).ElementId
				dbGraph.EdgeIds[edge] = id
				dbGraph.Edges[id] = edge
			}
```

Running with Neo4j 4:
Insert 259 relationships using query:
```
unwind $edges as edge 
match(a),(b) where id(a)=edge.from and id(b)=edge.to 
create (a)-[e:`has`]->(b) set e=edge.props return e
```
7.991203ms 
Collecting the results takes: 2.433761ms

Same code/data running with Neo4j 5:

Query:
```
unwind $edges as edge 
match(a),(b) where elementId(a)=edge.from and elementId(b)=edge.to 
create (a)-[e:`has`]->(b) set e=edge.props return e
```

1.682275ms 
Collect the results takes: 2.617625869s


When I change the query to:
```
unwind $edges as edge 
match(a),(b) where elementId(a) in [edge.from] and elementId(b) in [edge.to] 
create (a)-[e:`has`]->(b) set e=edge.props return e
```
Insert 259 edges:  1.355121ms 
Collect 259 edges: 13.740233ms


So using `IN` changes things a lot.

**Comment by inmost-light** [comment]
@bserdar 
> So using `IN` changes things a lot.

You should see `NodeByElementIdSeek` in the `EXPLAIN` output then, which would explain the better running time.

We've implemented a planner fix which now plans `NodeByElementIdSeek` for both your original query and the `IN` version. The fix should be included in Neo4j 5.6.0.

Does this resolve the issue for you?

**Comment by bserdar** [comment]
Yes, this resolves the issue, thank you.
